src/redist/targets/GenerateMSIs.targets

```diff
@@ -127,28 +135,61 @@
                       -InformationAction Continue" />
   </Target>
 
-  <Target Name="GenerateTemplatesMsi"
-        DependsOnTargets="GenerateLayout;AcquireWix;MsiTargetsSetupInputOutputs;SetSdkBrandingInfo"
-        Condition=" '$(OS)' == 'Windows_NT' "
-        Inputs="$(TemplatesLayoutPath)**/*;
-                  $(TemplatesGenerateMsiPowershellScript)"
-        Outputs="$(TemplatesMSIInstallerFile)">
+  <Target Name="GenerateTemplatesMsis"
+          DependsOnTargets="GenerateLayout;AcquireWix;MsiTargetsSetupInputOutputs;SetSdkBrandingInfo;SetupTemplatesMsis"
+          Condition="$(ProductMonikerRid.StartsWith('win')) And !$(Architecture.StartsWith('arm'))"
+          Inputs="@(TemplatesMsiComponent->'%(LayoutPath)');
+                    $(TemplatesGenerateMsiPowershellScript)"
+          Outputs="@(TemplatesMsiComponent->'%(MSIInstallerFile)')">
 
     <Exec Command="powershell -NoProfile -NoLogo $(TemplatesGenerateMsiPowershellScript) ^
-                      '$(TemplatesLayoutPath.TrimEnd('\'))' ^
-                      '$(TemplatesMSIInstallerFile)' ^
+                      '@(TemplatesMsiComponent->'%(LayoutPath)')' ^
+                      '%(TemplatesMsiComponent.MSIInstallerFile)' ^
```

Comment:
I am not 100% sure this syntax here is equivalent to the one you are using above and below.

Reply:
After checking the 3.0 pass - the syntax looks equivalent:
@(TemplatesMsiComponent->'%(LayoutPath)') evaluates to: $(Templates30LayoutPath.TrimEnd('\'))
%(TemplatesMsiComponent.MSIInstallerFile) evaluates to: $(Templates30MSIInstallerFile)

Reply:
Here's my understanding. @rainersigwald to correct me if I get it wrong.

One of them is a [transform](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-transforms?view=vs-2019), the other is [batching](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching?view=vs-2019).

The transform turns the list of TemplatesMsiComponent into the list of TemplatesMsiComponent.LayoutPath.  This is like Select in LINQ or map in other languages. The batching causes the Exec task to be run multiple times, passing each TemplatesMsiComponent.MSIInstallerFile.

Now here's the tricky part. The batching actually influences the transform. The TemplatesMsiComponent list is reduced to the TemplatesMsiComponent in the current batch iteration. There will be one per iteration, so transforming the single item is not really different than another %(A.B) reference.

I recommend replacing the transform with %(TemplatesMsiComponent.LayoutPath).That would be easier to read and understand.

Reply:
Changed.